designs/2023-prelints/README.md

```diff
@@ -0,0 +1,452 @@
+];
+```
+
+Here, the `"markdown/js"` prelint is a prelint that will extract JavaScript code from Markdown files and `"markdown/css"` is a prelint that will extract CSS code from Markdown files.
+
+Prelints can be configured similar to rules, except prelints will accept any truthy value as "on" and any falsy value as "off". There is no need for severities because prelints do not report violations. If a prelint supports options, it can be passed in config as well:
+
+```js
+// inside eslint.config.js
+import markdown from "eslint-plugin-markdown";
+
+export default [
+    {
+        plugins: {
+            markdown
+        },
+        files: ["**/*.md"],
+        language: "markdown/markdown",
+        prelints: {
+
+            // prelint with options
+            "markdown/js": {
+                codeBlockTags: ["js", "javascript", "ecmascript"]
+            },
+            "markdown/css": true
+        }
+    }
+];
+```
+
+Here, the `"markdown/js"` prelint is being passed options it can use to modify its behavior.
+
+### Prelint Definition Object
+
+Each prelint definition object must implement the following interface:
+
+```ts
+interface Prelint {
+
+    /**
+     * Space for any needed meta information.
+     */
+    meta: {
+        docs: {
+            description: string;
+        }
+    }
+
+    /**
+     * Creates the prelint.
+     */
+    create(context: PrelintContext): Object;
+
+}
+
+// Supplemental interfaces
+
+interface PrelintContext {
+    /**
+     * Options passed in config.
+     */
+    options: Array<any> | Object;
+
+    /**
+     * The SourceCode object to work on.
+     */
+    sourceCode: sourceCode;
+
+    /**
+     * Shared settings from config.
+     */
+    settings: object;
+
+    /**
+     * ID of the prelint (i.e., "markdown/js")
+     */
+    id: string;
+
+    /**
+     * Current working directory.
+     */
+    cwd: string;
+
+    /**
+     * Filename that ESLint is working on, including code blocks.
+     */
+    filename: string;
+
+    /**
+     * Physical filename ESLint is working on, excluding code blocks.
+     */
+    physicalFilename: string;
+
+
+    /**
+     * Creates a virtual file from a subset of the current file.
+     */
+    createVirtualFile(file: VirtualFile): void;
+}
+
+// Supplemental interfaces
+
+type VirtualFile = VirtualTextFile | VirtualBinaryFile;
+
+interface VirtualTextFile {
+
+    /**
+     * The name of the file without a path.
+     */
+    filename: string;
+
+    /**
+     * The body of the file to parse.
+     */
+    body: string | ArrayBuffer;
+
+    /**
+     * The number to add to any line-based violations. 0 if undefined.
+     */
+    lineStart?: number;
+
+    /**
+     * The number to add to any violations that occur in the first line of
+     * the virtual file. 0 if undefined.
+     */
+    columnStart?: number;
+
+    /**
+     * The number to add to the column number of each violation in the body.
+     * 0 if undefined.
+     */
+    indentOffset?: number;
```

Comment:
Is there a way to tell the ESLint engine the negative column offset of a line of text?

For example, if I have HTML like the following (for a 2-space indent), if I want the `indentOffset` to be `4`, the `font-size` will be placed at a column offset of `-4`.

```html
<div>
   <div style="
     color: red;
font-size: 16px;
   ">
```

Reply:
The reason I want to use 4 for indentOffset is because I want to convey the base position for the indent rule. If it's 0, I don't think the indent rule work well. It's probably indented like this:

```html
<div>
  <div style="
color: red;
font-size: 16px;
  ">
```

I think people actually want inden like this:

```html
<div>
  <div style="
    color: red;
    font-size: 16px;
  ">
```

Reply:
> Is there a way to tell the ESLint engine the negative column offset of a line of text?

Hmmm, I'm not sure I understand why you'd want to do this. If you have this HTML:

```html
<div>
  <div style="
    color: red;
font-size: 16px;
  ">
```

And you want to extract the CSS, for linting, I'd think you'd want ESLint to see the CSS as this:

```css
    color: red;
font-size: 16px;
```

So you'd say:

```js
context.createVirtualFile({
    body: removeIndents(cssNode.text),    // get the text you want somehow
    lineStart: 3,    // add this number to any reported lines
    columnStart: 0,    // add this number to any columns reported in the first line
    indentOffset: 0    // don't adjust indents
});
```

So then you'd take the CSS virtual file, normalize the indents, then that would get merged back into the HTML.

Would that work?

Reply:
Sorry, I think I didn't explain well 😓. (Prelint may be an unrelated topic.)
How do you think indentation rules for CSS should be implemented to handle indentation correctly given text like this?

```html
<div>
  <div style="
    color: red;
font-size: 16px;
  ">
```

I would expect it to be indented like this:

```html
<div>
  <div style="
    color: red;
    font-size: 16px;
  ">
```

However, I think the current system would probably indent like this:

```html
<div>
  <div style="
color: red;
font-size: 16px;
  ">
```

Because the CSS indentation rules don't know the proper starting position.

I thought it could be solved by indenting based on 0 if a negative column offset could be used, but there may be another way to solve it.

Reply:
Ah I see what you mean. Indeed, the current design doesn't allow for that. I need to think a bit more about how to handle this.

I think negative indents would be a bit confusing, as a negative indent would mean "add indents" while a positive indent would mean "remove indents". 

It might be the case that we need some sort of callback function to allow prelints to further modify a fixed result before inserting into the original file.

I'll think more on this and see what I can come up with.

Reply:
Okay, so it turns out this will work, just not in the most obvious way. 

First, we actually need two different rules, one for CSS and one for HTML. Let's call them `css/indent` to indent CSS and `html/indent` to indent HTML.

So to start, we have this code:

```html
<div>
  <div style="
    color: red;
font-size: 16px;
  ">
```

The prelint would extract the CSS into a text fragment that looks like this:

```css
    color: red;
font-size: 16px;
```

The `css/indent` rule would fix the indentation to this:

```css
color: red;
font-size: 16px;
```

Then, this CSS would be inserted back into the HTML, looking like this:

```html
<div>
  <div style="
color: red;
font-size: 16px;
  ">
```

When the `html/indent` rule then runs, it would end up fixing to this:

```html
<div>
  <div style="
    color: red;
    font-size: 16px;
  ">
```

Basically, we'd be combining the fixes from `css/indent` and `html/indent` together either in one pass (if technically possible) or at most in two passes (ESLint does multiple passes when attempting to autofix, so this is not new).